### PR TITLE
feat: adds EnabledWhen to the feature builder

### PR DESCRIPTION
**Pull request by cam-garrison**
## Description

This PR introduces a featureBuilder function that enables conditional toggling of features named `EnabledWhen`. Features can now be activated or deactivated based on specific conditions such as the target platform, management mode, or the presence of particular resources in the cluster.

Building on the previous PR linked below which we didn't finish up, this update ensures that feature.Cleanup() is called if a feature's EnabledWhen() status changes from true to false across multiple reconciles. Additionally, this PR includes enhanced testing to validate these changes.

## JIRA issue:

https://issues.redhat.com/browse/RHOAIENG-8540

re-doing and finishing up work from PR #793 

## How Has This Been Tested?

Wrote additional cleanup integration tests - using an example function checking if a certain namespace exists to enable the feature. 

Ensured the feature applied when the namespace exists and that the feature was cleaned up on "reconcile" when the namespace no longer exists. 

## Merge criteria:

- [x] Have a meaningful commit messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by zdtsw** [comment]
so this EnabledWhen can be used to replace PreCondition in some way?

**Comment by cam-garrison** [comment]
> so this EnabledWhen can be used to replace PreCondition in some way?

`PreCondition` will fail when trying to apply Features and `EnabledWhen` will only apply feature when condition is met. So we could in some cases end up replacing with `EnabledWhen` (ie authorino exists, install xyz) but the two functionalities are not interchangeable.

**Review by zdtsw — APPROVED** [review]

**Comment by cam-garrison** [comment]
added hold label to get additional review from @bartoszmajsak before allowing auto-merge

File: tests/integration/features/cleanup_int_test.go
Comment (bartoszmajsak) [inline tests/integration/features/cleanup_int_test.go:62]: Any reason to remove secret name as a param? With this change we hardcode the secret name in the func itself, not sure we want that.

File: tests/integration/features/cleanup_int_test.go
Comment (bartoszmajsak) [inline tests/integration/features/cleanup_int_test.go:155]: I wonder if we shouldn't test something different here. We should ensure nothing that is supposed to be created by the defined feature (including tracker) should exist in the cluster.

On top of that, `Consistently` might be better instead of `Eventually` (which can pass in the first run). The clean-up from the previous run might not be done yet on the cluster when this test is executed. I see you are handling it with the current implementation by checking if ns is in a deletion state, though I believe switching it to `Consistently` should make it safer if we happen to rework the test fixture, but keep the expectations the same.

File: tests/integration/features/cleanup_int_test.go
Comment (bartoszmajsak) [inline tests/integration/features/cleanup_int_test.go:163]: In `createdSecretHasOwnerReferenceToOwningFeature` shouldn't we ensure we look up the FT matching the feature, not assuming there's a FT matching any Feature in the cluster that was related to "test-secret"? In other words I would think this function also expects a Feature to be passed by only reading the signature. Otherwise we assume a bit too much internally.

File: tests/integration/features/cleanup_int_test.go
Comment (bartoszmajsak) [inline tests/integration/features/cleanup_int_test.go:201]: From the godoc: 

```golang
// IsNotFound returns true if the specified error was created by NewNotFound.
// It supports wrapped errors and returns false when the error is nil.
func IsNotFound(err error) bool 
```

so with that in mind, this should be sufficient:

```suggestion
	if errors.IsNotFound(err) {
		return false
	}
```

File: pkg/feature/builder.go
Comment (bartoszmajsak) [inline pkg/feature/builder.go:129]:
```suggestion
// EnabledWhen determines if a Feature should be loaded and applied based on specified criteria. 
// The criteria are supplied as a function.
//
// Note: The function passed should consistently return true while the feature is needed. 
// If the function returns false at any point, the feature's contents might be removed during the reconciliation process.
```

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]

File: tests/integration/features/cleanup_int_test.go
Comment (cam-garrison) [reply tests/integration/features/cleanup_int_test.go:62]: This was removed as linter complained as we were passing in the same secret name each time IIRC

**Comment by cam-garrison** [comment]
linter fails when nolint has preceding space and fails when it does not have preceding space : / fixing this tmrw

File: pkg/feature/builder.go
Comment (bartoszmajsak) [reply pkg/feature/builder.go]: From offline discussion, the flow here was wrong, that should be:

```golang
	// If the feature is disabled, but the FeatureTracker exists in the cluster, ensure clean-up is triggered.
	// This means that the feature was previously enabled, but now it is not anymore.
	if !feature.Enabled {
		if errGet := getFeatureTrackerIfAbsent(feature); err == nil {
			return feature.Cleanup(ctx)
		} 
		
		return client.IgnoreNotFound(err)
	}
```

**Review by bartoszmajsak — APPROVED** [review]
LGTM

**Pull request merged**
